proc.c

```diff
@@ -680,6 +680,29 @@ cfunc_proc_new(VALUE klass, VALUE ifunc)
     return procval;
 }
 
+VALUE
+rb_func_proc_dup(VALUE src_obj)
+{
+    RUBY_ASSERT(rb_typeddata_is_instance_of(src_obj, &proc_data_type));
+
+    rb_proc_t *src_proc;
+    GetProcPtr(src_obj, src_proc);
+    RUBY_ASSERT(vm_block_type(&src_proc->block) == block_type_ifunc);
+
+    cfunc_proc_t *proc;
+    VALUE proc_obj = TypedData_Make_Struct(rb_obj_class(src_obj), cfunc_proc_t, &proc_data_type, proc);
+
+    memcpy(&proc->basic, src_proc, sizeof(rb_proc_t));
```

Comment:
```suggestion
    proc->basic = src_proc->basic;
```

Tricky memcpy assuming layout doesn't seem necessary here.

Reply:
We can't do that because the `block` in `rb_proc_t` is defined as a const.

Reply:
```suggestion
    memcpy(&proc->basic, &src_proc->basic, sizeof(proc->basic));
```
Ah okay, let's just make lhs the same as rhs then.

Reply:
😅 Unfortunately the LHS and RHS are different types. LHS is a `cfunc_proc_t` while RHS is a `rb_proc_t`.